idaes/apps/grid_integration/multiperiod/price_taker_model.py

```diff
@@ -0,0 +1,429 @@
+
+        Returns:
+            daily_data: reconfigured price series data
+        """
+        # Get column headings
+        column_head = raw_data.columns.tolist()
+        # Remove the date/time column
+        scenarios = column_head[2:]
+
+        # Creating an empty dataframe to store daily data for clustering
+        day_list = list(range(1, (len(raw_data) // self._horizon_length) + 1))
+
+        # Generate daily data
+        for i in scenarios[0:1]:
+            daily_data = self.generate_daily_data(
+                raw_data=raw_data[i], day_list=day_list
+            )
+
+        return daily_data, scenarios
+
+    def get_optimal_n_clusters(self, daily_data, kmin=None, kmax=None, plot=False):
+        """
+        Determines the appropriate number of clusters needed for a
+        given price signal.
+
+        Args:
+            daily_data: reconfigured price series data
+            kmin: minimum number of clusters
+            kmax: maximum number of clusters
+            plot: flag to determine if an elbow plot should be displayed
+
+        Returns:
+            n_clusters: the optimal number of clusters for the given data
+            inertia_values: within-cluster sum-of-squares
+        """
+        if kmin is None:
+            kmin = 4
+        if kmax is None:
+            kmax = 30
+            _logger.warning(f"kmax was not set - using a default value of 30.")
+
+        k_values = range(kmin, kmax)
+        inertia_values = []
+
+        np.random.seed(self._seed)
+
+        # Compute the inertia (SSE) for k clusters
+
+        for k in k_values:
+            kmeans = KMeans(n_clusters=k).fit(daily_data.transpose())
+            inertia_values.append(kmeans.inertia_)
+
+        # Identify the "elbow point"
+        elbow_point = KneeLocator(
+            k_values, inertia_values, curve="convex", direction="decreasing"
+        )
+        n_clusters = elbow_point.knee
+
+        print(f"Optimal # of clusters is: {n_clusters}")
+
+        if n_clusters + 2 >= kmax:
+            _logger.warning(
+                f"Optimal number of clusters is close to kmax: {kmax}. Consider increasing kmax."
+            )
+
+        if plot == True:
+            plt.show()
+            plt.plot(k_values, inertia_values)
+            plt.axvline(x=n_clusters, color="red", linestyle="--", label="Elbow")
+            plt.xlabel("Number of clusters")
+            plt.ylabel("Inertia")
+            plt.title("Elbow Method")
+            plt.xlim(kmin, kmax)
+            plt.grid()
+            plt.show()
+
+        return n_clusters, inertia_values
+
+    def cluster_lmp_data(self, daily_data, n_clusters, scenarios):
+        """
+        Clusters the given price signal in n_clusters. This method supports k-means, k-meteiod,...
+        techniques for clustering.
+
+        Args:
+
+        Returns:
+            lmp_data = {1: {1: 2, 2: 3, 3: 5}, 2: {1: 2, 2: 3, 3: 5}}
+            weights = {1: 45, 2: 56}
```

Comment:
Currently the output for weights is in the format {0: {0 45 , 1: 56, ...). I assume we want to get rid of the 0 in front so that it matches the format Radha initially put in the doc description, but I'm not sure how to go about doing that. I _think_ it's just the header for the column similar to how the output for lmp_data is in the format: {0: {1: 2, 2: 3, 3: 5}, 1: {1: 2, 2: 3, 3: 5}}, where 0 is the header of the first column representing the first cluster. 

On that last note, is it okay if `lmp_data` and `weights` start at 0 rather than 1?